File: Classes/FrameBuffer.py
from collections import deque

class FrameBuffer:

    def __init__(self, size):
        self.buffer = deque(maxlen=size)

    def __iter__(self):
        return iter(self.buffer)

    def addFrame(self, frame):
        self.buffer.appendleft(frame)

    def getFrames(self):
        return list(self.buffer)
    
    def __getitem__(self, item):
        if (0 <= item and item < len(self.buffer)):
            return self.buffer[item]
        else:
            raise Exception("index given was out of range of FrameBuffer")
        
    def filterNilBasketballs(self, howManyFrames):
        nonNilElements = [frame for frame in self.buffer if frame.ball is not None]
        return nonNilElements[:howManyFrames]
        
    
        
    

File: BasketballLab/App/MainApp/Workouts,Camera/Camera/VisionDataModels/FrameBuffer.swift
//
//  FrameBuffer.swift
//  BasketballLab
//
//  Created by Onik Hoque on 2/27/24.
//

import Foundation
import DequeModule

struct FrameBuffer {
    
    var buffer : Deque<Frame>
    let maxLength : Int
    
    init(maxLength: Int) {
        self.maxLength = maxLength
        self.buffer = []
    }
    
    mutating func addLocationFrame(frame: Frame) {
        if (buffer.count >= self.maxLength) {
            self.buffer.removeLast()
        }
        self.buffer.prepend(frame)
    }
    
    subscript(index: Int) -> Frame {
            if index >= 0 && index < buffer.count {
                return self.buffer[index]
            } else {
                fatalError("index given was out of range of FrameBuffer")
            }
        }
    
    func nonNilFrames() -> Deque<Frame> {
        let filteredArray = self.buffer.filter { $0.ball != nil }
        return filteredArray
    }
    
}
